rust-runtime/aws-smithy-http-server-python/src/types.rs

```diff
@@ -431,6 +433,85 @@ impl ByteStream {
     }
 }
 
+/// Python Wrapper for [aws_smithy_types::Document].
+#[derive(Debug, Clone, PartialEq)]
+pub struct Document(aws_smithy_types::Document);
+
+impl IntoPy<PyObject> for Document {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        use aws_smithy_types::{Document as D, Number};
+
+        match self.0 {
+            D::Object(obj) => obj
+                .into_iter()
+                .map(|(k, v)| (k, Document(v).into_py(py)))
+                .collect::<HashMap<_, _>>()
+                .into_py(py),
+            D::Array(vec) => vec
+                .into_iter()
+                .map(|d| Document(d).into_py(py))
+                .collect::<Vec<_>>()
+                .into_py(py),
+            D::Number(Number::Float(f)) => f.into_py(py),
+            D::Number(Number::PosInt(pi)) => pi.into_py(py),
+            D::Number(Number::NegInt(ni)) => ni.into_py(py),
+            D::String(str) => str.into_py(py),
+            D::Bool(bool) => bool.into_py(py),
+            D::Null => py.None(),
+        }
+    }
+}
+
+impl FromPyObject<'_> for Document {
+    fn extract(obj: &PyAny) -> PyResult<Self> {
+        use aws_smithy_types::{Document as D, Number};
+
+        if let Ok(obj) = obj.extract::<HashMap<String, Document>>() {
+            Ok(Self(D::Object(
+                obj.into_iter().map(|(k, v)| (k, v.0)).collect(),
+            )))
+        } else if let Ok(vec) = obj.extract::<Vec<Self>>() {
+            Ok(Self(D::Array(vec.into_iter().map(|d| d.0).collect())))
+        } else if let Ok(b) = obj.extract::<bool>() {
+            // This check must happen before any number checks because they cast
+            // `true`, `false` to `1`, `0` respectively.
+            Ok(Self(D::Bool(b)))
+        } else if let Ok(pi) = obj.extract::<u64>() {
+            Ok(Self(D::Number(Number::PosInt(pi))))
+        } else if let Ok(ni) = obj.extract::<i64>() {
+            Ok(Self(D::Number(Number::NegInt(ni))))
+        } else if let Ok(f) = obj.extract::<f64>() {
+            Ok(Self(D::Number(Number::Float(f))))
+        } else if let Ok(s) = obj.extract::<String>() {
+            Ok(Self(D::String(s)))
+        } else if obj.is_none() {
+            Ok(Self(D::Null))
+        } else {
+            Err(PyTypeError::new_err(format!(
+                "'{obj}' cannot be converted to 'Document'",
+            )))
+        }
+    }
+}
+
+// TODO(PythonSerialization): Get rid of this hack.
+// `JsonValueWriter::document` expects `&aws_smithy_types::Document`
+// and this impl allows `&Document` to get coerced to `&aws_smithy_types::Document`.
+// We should ideally handle this in `JsonSerializerGenerator.kt` but I'm not sure how hard it is.
```

Comment:
This is the same for also the rest of the types in this file. I think handling this in the JsonSerializer is hard as I was not able to find the right customization entrypoints at the time.